// @nrwl/nx/fesm5/nrwl-nx-testing.js
import { first, toArray } from 'rxjs/operators';
export { cold, hot, getTestScheduler, time } from 'jasmine-marbles';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/**
 * \@whatItDoes reads all the values from an observable and returns a promise
 * with an array of all values. This should be used in combination with async/await.
 *
 * ## Example
 *
 * ```typescript
 * const obs = of(1, 2, 3, 4);
 * const res = await readAll(obs)
 * expect(res).toEqual([1, 2, 3, 4]);
 * ```
 * @template T
 * @param {?} o
 * @return {?}
 */
function readAll(o) {
    return o.pipe(toArray()).toPromise();
}
/**
 * \@whatItDoes reads the first value from an observable and returns a promise
 * with it. This should be used in combination with async/await.
 *
 * ## Example
 *
 * ```typescript
 * const obs = of(1, 2, 3, 4);
 * const res = await readFirst(obs)
 * expect(res).toEqual(1);
 * ```
 * @template T
 * @param {?} o
 * @return {?}
 */
function readFirst(o) {
    return o.pipe(first()).toPromise();
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

export { readAll, readFirst };
