// @nrwl/nx/bundles/nrwl-nx-testing.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('rxjs/operators'), require('jasmine-marbles')) :
    typeof define === 'function' && define.amd ? define('@nrwl/nx/testing', ['exports', 'rxjs/operators', 'jasmine-marbles'], factory) :
    (factory((global.nrwl = global.nrwl || {}, global.nrwl.nx = global.nrwl.nx || {}, global.nrwl.nx.testing = {}),global.rxjs.operators,global['jasmine-marbles']));
}(this, (function (exports,operators,jasmineMarbles) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    /**
     * \@whatItDoes reads all the values from an observable and returns a promise
     * with an array of all values. This should be used in combination with async/await.
     *
     * ## Example
     *
     * ```typescript
     * const obs = of(1, 2, 3, 4);
     * const res = await readAll(obs)
     * expect(res).toEqual([1, 2, 3, 4]);
     * ```
     * @template T
     * @param {?} o
     * @return {?}
     */
    function readAll(o) {
        return o.pipe(operators.toArray()).toPromise();
    }
    /**
     * \@whatItDoes reads the first value from an observable and returns a promise
     * with it. This should be used in combination with async/await.
     *
     * ## Example
     *
     * ```typescript
     * const obs = of(1, 2, 3, 4);
     * const res = await readFirst(obs)
     * expect(res).toEqual(1);
     * ```
     * @template T
     * @param {?} o
     * @return {?}
     */
    function readFirst(o) {
        return o.pipe(operators.first()).toPromise();
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */

    exports.cold = jasmineMarbles.cold;
    exports.hot = jasmineMarbles.hot;
    exports.getTestScheduler = jasmineMarbles.getTestScheduler;
    exports.time = jasmineMarbles.time;
    exports.readAll = readAll;
    exports.readFirst = readFirst;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
